engine/crates/engine-v2/src/sources/graphql/request/prepare.rs

```diff
@@ -196,23 +258,93 @@ impl QueryBuilderContext {
             .fields_ordered_by_parent_entity_id_then_position()
             .into_iter()
             .chunk_by(|field| field.definition().parent_entity_id);
+
         for (entity_id, fields) in entity_to_fields.into_iter() {
-            if maybe_entity_id != Some(entity_id) {
-                indent_write!(
-                    buffer,
-                    "... on {} {{\n",
-                    selection_set.walker().schema().walk(entity_id).name()
-                )?;
-                buffer.indent += 1;
-            }
-            for field in fields {
-                self.write_field(buffer, field)?;
-            }
-            if maybe_entity_id != Some(entity_id) {
-                buffer.indent -= 1;
-                indent_write!(buffer, "}}\n")?;
+            let fields = fields.collect_vec();
+            let entity = selection_set.walker().schema().walk(entity_id);
+            let in_same_entity = maybe_entity_id == Some(entity_id);
+
+            let rendering = SelectionSetRendering::new(
+                maybe_entity_id,
+                maybe_selection_set_type,
+                entity,
+                self.subgraph_id,
+                selection_set,
+            );
+
+            match rendering {
+                SelectionSetRendering::InterfaceWithPartialFragment(interface, subgraph_id) => {
+                    for r#type in interface.possible_types_ordered_by_typename() {
+                        if !r#type.subgraph_implements_interface(&subgraph_id, &interface.id()) {
+                            continue;
+                        }
+
+                        self.write_type_fields(buffer, r#type.name(), &fields)?;
+                    }
+                }
+                SelectionSetRendering::UnionWithPartialFragment(union) => {
+                    for r#type in union.possible_types_ordered_by_typename() {
+                        self.write_type_fields(buffer, r#type.name(), &fields)?;
```

Comment:
Here it's also broken and hidden by the logic being separated. This case is parent is an union and child an interface. So you should end up  with the same code as before if the interface isn't fully implemented in the subgraph